torch/ao/quantization/pt2e/port_metadata_pass.py

```diff
@@ -0,0 +1,215 @@
+    if len(node_users) != 1:
+        raise InternalError(f"Expecting {node} to have single user")
+    q_node = node_users.pop()
+    if q_node.op != "call_function" or q_node.target not in _QUANTIZE_OPS:
+        logger.warn(f"Expecting {node} user to be a quantized op but got {q_node}")
+        return
+
+    _add_metadata(q_node, node)
+
+
+def _is_valid_annotation(annotation: QuantizationAnnotation) -> bool:
+    if annotation is None:
+        return False
+    input_qspec_map = annotation.input_qspec_map
+    output_qspec = annotation.output_qspec
+    if len(input_qspec_map) == 0 and output_qspec is None:
+        return False
+    return True
+
+
+class PortNodeMetaForQDQ(_ExportPassBase):
+    """
+    Port metadata for nodes added by quantization flow.
+    For static quant these are:
+    - quantizer_per_tensor.default, dequantize_per_tensor.default
+    - quantizer_per_channel.default, dequantize_per_channel.default
+    For dynamic quant these are:
+    - choose_qparams.tensor
+    - quantizer_per_tensor.tensor, dequantize_per_tensor.tensor
+    - quantizer_per_channel.default, dequantize_per_channel.default
+
+    Rules of porting metadata:
+    - Metadata to be ported:
+      - nn_module_stack
+      - stack_trace
+      - quantization_tag
+    - Metadata to NOT be ported:
+      - Everything else
+    - Rules:
+      - Statically quantized patterns:
+        - Dequantize nodes on the inputs to be quantized inherit metadata of the consumer node.
+        - Quantize nodes on the outputs inherit metadata of the producer node.
+        - For example:
+          - Original: [Conv -> AvgPool -> Linear]
+          - Quantized [Q-> DQ -> Conv -> Q -> DQ -> AvgPool -> Q -> DQ -> Linear -> Q -> DQ]
+          - Inner brackets specify which nodes Q/DQ inherit metdata from
+          - [Q-> [DQ -> Conv -> Q] -> [DQ -> AvgPool -> Q] -> [DQ -> Linear -> Q] -> DQ]
+          - Note first Q and last DQ do not inherit metadata from any nodes
+      - Dynamically quantized patterns:
+        - Input that are dynamically quantized have chose_qparams, quantize and dequantize nodes
+        - For example, below linear is dynamically quantized while rest statically:
+          - Original: [Conv -> AvgPool -> Linear]
+          - Quantized [Q-> DQ -> Conv -> Q -> DQ -> AvgPool -> Q -> DQ -> chose_params -> Q -> DQ -> Linear]
+          - Quantized [Q-> [DQ -> Conv -> Q] -> [DQ -> AvgPool -> Q] -> DQ -> [chose_params -> Q -> DQ -> Linear]]
+          - Note first Q does not inherit metadata from any nodes
+    """
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        for node in graph_module.graph.nodes:
+            annotation = node.meta.get("quantization_annotation", None)
+            if _is_valid_annotation(annotation):
+                input_qspec_map = node.meta["quantization_annotation"].input_qspec_map
+                output_qspec = node.meta["quantization_annotation"].output_qspec
```

Comment:
do we need to care about annotation in convert step?

Reply:
I think the reason annotation is inspected is to that you can distinguish between these uses cases

conv -> add -> linear: quantize with all 3 nodes annotated, you get
q -> dq -> conv -> q -> dq -> add -> q -> dq -> linear -> q -> dq. here quantized op patterns are in brackets
q -> [dq -> conv -> q] -> [dq -> add -> q] -> [dq -> linear -> q] -> dq

While only quantizing conv and linear, i.e. no annotation on add you get
q -> dq -> conv -> q -> dq -> add -> q -> dq -> linear -> q -> dq: quantized op patterns
q -> [dq -> conv -> q] -> dq -> add -> q -> [dq -> linear -> q] -> dq


Both graphs are exactly the same but quantized op patterns are different because in the latter case you dont want add to be quantized. Without inspecting quantization annotation, we cannot know if add was meant to be quantized or not.

Reply:
I see, is the goal here for annotation to control lowering? I feel annotation should not be coupled with lowering, can we just have another api in lowering step for lowering users to control how they want their model to be lowered?

Reply:
like you mentioned before, people can choose to use or not use these tags for lowering, these just expresses what quant flow counts as quantized op, since even with the same tags, people can choose to lower differently

Reply:
another thing I'm thinking of is that I think we should have each step self contained:

1. original model + quantizer
2. annotated model (now the information from quantizer should be 100% transferred to annotation
3. prepared model (now annotation information should be 100% transferred to prepared model, to observer/fake quants)
4. converted model (now prepared model information should be 100% transferred to converted model)

otherwise there will be extra burdens added to serialization I think. for example if in step 4. we expect quantization annotation to be there, then that means the quantization annotation need to be serialized in prepared model as well, and the prepared model is not self contained by itself.

Reply:
> I see, is the goal here for annotation to control lowering? I feel annotation should not be coupled with lowering, can we just have another api in lowering step for lowering users to control how they want their model to be lowered?

Annotation is expressing the intent of what is to be quantized. And if this intent is not communicable to lowering stage, it is a problem.

> like you mentioned before, people can choose to use or not use these tags for lowering, these just expresses what quant flow counts as quantized op, since even with the same tags, people can choose to lower differently

Yes, but with porting of tags it makes it much easier to decode user's intent of what was to be quantized and what was not. I can imagine during delegation, the tags are used to figure out what would be run on cpu and what on special backend.

Reply:
> otherwise there will be extra burdens added to serialization I think. for example if in step 4. we expect quantization annotation to be there, then that means the quantization annotation need to be serialized in prepared model as well, and the prepared model is not self contained by itself.

That is true. So the question is on serialization and what part of meta will be serialized or not.

Alternatively you have to capture the intent, as expressed by annotations, in the IR itself (if meta is not considrered part of IR). But not that even FX QAT today requires you to prepare the model again and if you chose and different way of doing prepare then there is a loss. So I dont think it is regressing from that behavior anyway. However, I do think preserving annotations somehow through serialization is important. Otherwise intent is lost. Now if that happens via meta serialization or otherwise, is upto what is possible.

Reply:
understood. I think we need to align a bit on the role of quantization_tags I think. right now my understanding is even without quantization_tag, user should be able to lower, since the graph is the ground truth and it should contain all the information they need for lowering. in that case quantization_tag is an optional utility that expresses quant_flows view on these ops. and user can choose not to follow the advice of quantization_tag and lower differently, e.g. they can lower [q -> dq -> conv -> q -> dq] to be a single op.

looks like you were saying it is a mandatory thing that all users should be aware of in order to express the intent?

Reply:
> should be aware of in order

I am not saying it is mandatory or not.

There are two possibilities
1. Lower using IR only. Loss of user's intent
2. Lower using IR + something else. QUantization_tag or annotations. Likely no loss of user's intent. quantization_tag allows you to do preserve users' intent, but is not the only way to do that. It just makes it bit more convenient.

Reply:
to clarify, what is the exact definition of user's intent? it seems to make perfect sense in the cases above and for the cases that user just want to lower [dq -> fp32_op -> q] patterns. but what about the pattern I mentioned [q -> dq -> conv -> q -> dq], people might annotate the conv node the exact same way as if they want to lower with [dq -> conv -> q] pattern, but how can they express the lowering intent of [q -> dq -> conv -> q -> dq]?

I feel there is intent of inserting q/dq ops, and intent of lowering. right now it looks like we are trying to express both in the annotation API and I feel not all lowering intent can be expressed with the current design and also feel it's cleaner to separate the inserting q/dq ops and lowering intent for a most general and extensible design.

Reply:
TBH I do feel it's easier UX if annotation is just expressing quantized ops directly, but seems like the reality is there is a separation between inserting q/dq ops and lowering and they are just separate, even though a lot of cases they are coupled

Reply:
> to clarify, what is the exact definition of user's intent? it seems to make perfect sense in the cases above and for the cases that user just want to lower [dq -> fp32_op -> q] patterns. but what about the pattern I mentioned [q -> dq -> conv -> q -> dq], people might annotate the conv node the exact same way as if they want to lower with [dq -> conv -> q] pattern, but how can they express the lowering intent of [q -> dq -> conv -> q -> dq]?

In case of "conv -> add -> linear", user doesnt want to quantize add but wants to quantize conv. This is not about lowering intent preservation as much as associating q/dq nodes. In quantizing conv->add->linear where resulting pattern is `q1 -> dq1 -> conv -> q2 -> dq2 -> add -> q3 -> dq3 -> linear -> q4 -> dq4`, dq2 and q3 are not associated with add because add was not be quantized. That is all this is about. SO I am not sure what part is not clear.

Reply:
might be easier to explain in a meeting, let's follow up